Fix blocking the bot

Fix the runtime error which occurs when the user blocks the bot:

```
TypeError: Cannot read property 'from' of undefined
    at onBlocked (/home/shronk/Programs/TypeScript/agora/telegram-runner/src/service/events.ts:86:38)
    at Object.onMyChatMemberUpdate (/home/shronk/Programs/TypeScript/agora/telegram-runner/src/service/events.ts:112:5)
    at /home/shronk/Programs/TypeScript/agora/telegram-runner/src/Bot.ts:39:48
    at /home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:158:111
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at execute (/home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:458:17)
    at /home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:459:21
    at execute (/home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:458:17)
    at /home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:459:21
    at execute (/home/shronk/Programs/TypeScript/agora/telegram-runner/node_modules/telegraf/lib/composer.js:458:17)
[ERROR] 11:13:55 TypeError: Cannot read property 'from' of undefined
```

--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     "dotenv": "^10.0.0",
     "express": "^4.17.1",
     "express-validator": "^6.11.1",
-    "telegraf": "^4.3.0",
+    "telegraf": "^4.4.0",
     "typegram": "^3.3.1",
     "winston": "^3.3.3"
   },

--- a/src/api/actions.ts
+++ b/src/api/actions.ts
@@ -69,9 +69,9 @@ const manageGroups = async (
         platformUserId,
         "You have 15 minutes to join these groups before the invite links " +
           "expire:",
-        Markup.inlineKeyboard([
-          invites.map((inv) => Markup.button.url(inv.name, inv.link))
-        ])
+        Markup.inlineKeyboard(
+          invites.map((inv) => [Markup.button.url(inv.name, inv.link)])
+        )
       );
     }
   } else {

--- a/src/api/actions.ts
+++ b/src/api/actions.ts
@@ -1,3 +1,4 @@
+import { Markup } from "telegraf";
 import Bot from "../Bot";
 import { ManageGroupsParam } from "./types";
 import logger from "../utils/logger";
@@ -29,24 +30,34 @@ const isMember = async (groupId: string, userId: number): Promise<Boolean> => {
   }
 };
 
+const getGroupName = async (groupId: string): Promise<string> => 
+   groupId // TODO: fix this
+;
+
 const manageGroups = async (
   params: ManageGroupsParam,
   isUpgrade: boolean
 ): Promise<boolean> => {
   const { platformUserId } = params;
-  console.log(isUpgrade, platformUserId, params);
+
   if (isUpgrade) {
-    const invites: string[] = [];
+    const invites: { link: string; name: string }[] = [];
+
     await Promise.all(
       params.groupIds.map(async (groupId) => {
         try {
-          const status = await isMember(groupId, +platformUserId);
-          if (!status) {
+          if (!(await isMember(groupId, +platformUserId))) {
             const inviteLink = await generateInvite(
               params.platformUserId,
               groupId
             );
-            if (inviteLink !== undefined) invites.push(inviteLink);
+
+            if (inviteLink !== undefined) {
+              invites.push({
+                link: inviteLink,
+                name: await getGroupName(groupId)
+              });
+            }
           }
         } catch (err) {
           logger.error(err);
@@ -55,21 +66,24 @@ const manageGroups = async (
     );
 
     if (invites.length) {
-      const message: string = `${
+      Bot.Client.sendMessage(
+        platformUserId,
         "You have 15 minutes to join these groups before the invite links " +
-        "expire: \n"
-      }${invites.join("\n")}`;
-
-      Bot.Client.sendMessage(platformUserId, message);
+          "expire:",
+        Markup.inlineKeyboard([
+          invites.map((inv) => Markup.button.url(inv.name, inv.link))
+        ])
+      );
     }
   } else {
     params.groupIds.forEach(async (groupId) => {
       try {
         const member = await Bot.Client.getChatMember(groupId, +platformUserId);
+
         if (member?.status === "member") {
-          Bot.Client.kickChatMember(groupId, +platformUserId).catch((e) =>
+          Bot.Client.kickChatMember(groupId, +platformUserId).catch(() =>
             logger.error(
-              `Couldn't remove Telegram user with userId "${platformUserId}"${e}`
+              `Couldn't remove Telegram user with userId "${platformUserId}"`
             )
           );
         }

--- a/src/api/actions.ts
+++ b/src/api/actions.ts
@@ -30,9 +30,8 @@ const isMember = async (groupId: string, userId: number): Promise<Boolean> => {
   }
 };
 
-const getGroupName = async (groupId: string): Promise<string> => 
-   groupId // TODO: fix this
-;
+const getGroupName = async (groupId: string): Promise<string> =>
+  ((await Bot.Client.getChat(groupId)) as { title: string }).title;
 
 const manageGroups = async (
   params: ManageGroupsParam,

--- a/src/service/commands.ts
+++ b/src/service/commands.ts
@@ -50,9 +50,9 @@ const listCommunitiesCommand = (ctx: any): void => {
   fetchCommunitiesOfUser(ctx.message.from.id).then((results) => {
     ctx.replyWithMarkdown(
       "Please visit your communities' websites:",
-      Markup.inlineKeyboard([
-        results.map((res) => Markup.button.url(res.name, res.url))
-      ])
+      Markup.inlineKeyboard(
+        results.map((res) => [Markup.button.url(res.name, res.url)])
+      )
     );
   });
 };

--- a/src/service/commands.ts
+++ b/src/service/commands.ts
@@ -48,12 +48,11 @@ const leaveCommand = (ctx: any): void => {
 
 const listCommunitiesCommand = (ctx: any): void => {
   fetchCommunitiesOfUser(ctx.message.from.id).then((results) => {
-    const urls = results
-      .map((result) => `[${result.name}](${result.url})`)
-      .join("\n");
-
     ctx.replyWithMarkdown(
-      `*Please visit your communities' websites:*\n${urls}`
+      "Please visit your communities' websites:",
+      Markup.inlineKeyboard([
+        results.map((res) => Markup.button.url(res.name, res.url))
+      ])
     );
   });
 };

--- a/src/service/events.ts
+++ b/src/service/events.ts
@@ -85,11 +85,15 @@ const onUserRemoved = (platformUserId: string, groupId: string): void => {
 const onBlocked = (ctx: any): void => {
   const platformUserId = ctx.update.my_chat_member.from.id;
 
-  fetchCommunitiesOfUser(platformUserId).then((communities) =>
-    communities.forEach((community) =>
-      leaveCommunity(platformUserId, community.id)
+  logger.warn(`User "${platformUserId}" has blocked the bot.`);
+
+  fetchCommunitiesOfUser(platformUserId)
+    .then((communities) =>
+      communities.forEach((community) =>
+        leaveCommunity(platformUserId, community.id)
+      )
     )
-  );
+    .catch(logger.error);
 };
 
 const onChatMemberUpdate = (ctx: any): void => {

--- a/src/service/events.ts
+++ b/src/service/events.ts
@@ -11,7 +11,7 @@ const onMessage = (ctx: any): void => {
       .then(() =>
         ctx.replyWithMarkdown(
           "You can find more information on the " +
-            "[Agora](https://agora.space/) website."
+            "[Agora](https://app.agora.space/) website."
         )
       );
   }

--- a/src/service/events.ts
+++ b/src/service/events.ts
@@ -105,9 +105,6 @@ const onChatMemberUpdate = (ctx: any): void => {
     logger.debug(invLink);
 
     onUserJoined(invLink, member.from.id, member.chat.id);
-
-    // TODO: check if the user fullfills the requirements
-    onUserRemoved(member.from.id, member.chat.id);
   }
 };
 

--- a/src/service/events.ts
+++ b/src/service/events.ts
@@ -83,7 +83,7 @@ const onUserRemoved = (platformUserId: string, groupId: string): void => {
 };
 
 const onBlocked = (ctx: any): void => {
-  const platformUserId = ctx.message.from.id;
+  const platformUserId = ctx.update.my_chat_member.from.id;
 
   fetchCommunitiesOfUser(platformUserId).then((communities) =>
     communities.forEach((community) =>

